proposals/0000-class-method-headers.rst

```diff
@@ -0,0 +1,437 @@
+
+  gendecl -> vars :: [context =>] type
+  vars    -> var_1 , ... , var_n         (n ≥ 1)
+
+Instead of variable names for class method left-hand sides, we introduce the
+notion of a signature header::
+
+  sighdr  -> var (sigbndr_1 ... sigbndr_n)
+  sigbndr -> tyvar
+           | @tyvar
+
+  gendecl -> sighdrs :: [context =>] type
+  sighdrs -> sighdr_1 , ... , sighdr_n         (n ≥ 1)
+
+A validity check ensures that the binders are only used in class method
+signatures and are disallowed in function signatures.
+
+The syntax of function bindings, including method definitions in instances, is
+assumed to be extended by `"Type variable binders"
+<https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0155-type-lambda.rst>`_
+to allow type binders prefixed with ``@``::
+
+  f @Int x = ...
+
+Semantics
+~~~~~~~~~
+
+1. When one or more binders are present in a class method signature, we consider
+   it a class method header, and require the ``-XClassMethodHeaders`` extension.
+
+#. A class method header, if present, must bind every class variable mentioned
+   in the class declaration header, and must bind it exactly once.
+
+#. No other variables can be bound in the class method header.
+
+#. In the compiler-generated top-level signature for the class method, variables
+   bound as ``@a`` are quantified with ``forall a.``, and variables bound as
+   ``a`` are quantified with ``forall a ->``.
+
+#. In the compiler-generated top-level signature for the class method, the order
+   of quantifiers matches the order of binders in the class method header.
+
+#. The variables may not be quantified out of dependency order.
+
+#. In instances, method definitions *may* explicitly write out instantiations of
+   invisible class variables::
+
+     class C a where
+       f @a :: a
+
+     instance C Int where
+       f @Int = 42
+
+   The type in the method equation must be identical to the one in the instance
+   head.
+
+#. In instances, method definitions *must* explicitly write out
+   instantiations of visible class variables::
+
+     class D a where
+       g a :: a
+
+     instance D Int where
+       g Int = 42
+
+   The type in the method equation must be identical to the one in the instance
+   head.
+
+#. The presence of a class method header is not observable outside the class
+   declaration, aside from the effect that it has on the order and visibility of
+   quantification::
+
+    class C a b where
+      f1       :: a -> b
+      f2 @a @b :: a -> b  -- fully identical
+
+   A consequence of this is that class method definitions may bind class
+   variables regardless of the presence of a class method header in the class
+   declaration::
+
+     class C a where
+       f :: a         -- note: no @a
+
+     -- f :: forall a. C a => a
+
+     instance C Int where
+       f @Int = 42    -- still allowed
+
+#. Class variables bound in methods always precede other arguments.
```

Comment:
This seems to answer my question above. I guess that turning this extension on can break programs which are implemented with `-XTypeAbstractions`, correct? If so, maybe a little word in the _Effect and interaction_ section would be warranted.